test/scale/scale_tester.go

```diff
@@ -0,0 +1,358 @@
+	klog.Infof("1 min: %d pods are ready", oneMinPodsRunning)
+	klog.Infof("2 min: %d pods are ready", twoMinPodsRunning)
+	klog.Infof("3 min: %d pods are ready", threeMinPodsRunning)
+	klog.Infof("5 min: %d pods are ready", fiveMinPodsRunning)
+	klog.Infof("10 min: %d pods are ready", tenMinPodsRunning)
+	klog.Infof("Total number of pods in Running state: %d", numberOfRunningPods)
+
+	klog.Infof("Scaling in test completed in %f minutes.", scaleInPodsCompleted)
+	klog.Infof("1 min: %d pods are deleted", oneMinPodsDeleted)
+	klog.Infof("2 min: %d pods are deleted", twoMinPodsDeleted)
+	klog.Infof("3 min: %d pods are deleted", threeMinPodsDeleted)
+	klog.Infof("5 min: %d pods are deleted", fiveMinPodsDeleted)
+	klog.Infof("10 min: %d pods are deleted", tenMinPodsDeleted)
+	klog.Infof("Total number of remaining pods: %d", numberOfRemainingPods)
+
+	klog.Infof("Scaling in detach completed in %f minutes.", scaleInCompleted)
+
+}
+
+func (t *PodSchedulingOnFailoverScaleTest) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	// Get the list of available nodes for scheduling the pod
+	nodes := nodeutil.ListNodeNames(client)
+	if len(nodes) < 3 {
+		ginkgo.Skip("need at least 3 nodes to verify the test case. Current node count is %d", len(nodes))
+	}
+
+	ginkgo.By("Initiating statefulset setup")
+
+	statefulSets := []*testtypes.TestStatefulset{}
+
+	tStorageClass, storageCleanup := t.Pod.CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer storageCleanup()
+
+	for i := 0; i < t.PodCount; i++ {
+		statefulSet, cleanup := t.Pod.SetupStatefulset(client, namespace, t.CSIDriver, schedulerName, 1, t.StorageClassParameters, &tStorageClass)
+		for j := range cleanup {
+			defer cleanup[j]()
+		}
+
+		statefulSets = append(statefulSets, statefulSet)
+	}
+
+	ginkgo.By("Completed statefulset setup")
+
+	ginkgo.By("Initiating statefulset deployment")
+	var wg sync.WaitGroup
+	start := time.Now()
+	for i := range statefulSets {
+		wg.Add(1)
+		go func(statefulset *testtypes.TestStatefulset) {
+			defer wg.Done()
+			defer ginkgo.GinkgoRecover()
+
+			statefulset.Create()
+		}(statefulSets[i])
+	}
+
+	ginkgo.By("Waiting for the pod for all the statefulsets to succeed")
+	wg.Wait()
+	framework.Logf("Time taken to complete deployment of statefulsets : %d", time.Since(start).Milliseconds())
+
+	ginkgo.By("All pods for the statefulsets are ready")
+	ginkgo.By("Initiating scaling down all the statefulsets")
+
+	for i := range statefulSets {
+		newScale := &scale.Scale{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      statefulSets[i].Statefulset.Name,
+				Namespace: statefulSets[i].Namespace.Name,
+			},
+			Spec: scale.ScaleSpec{
+				Replicas: int32(0),
+			},
+		}
+
+		_, err := client.AppsV1().StatefulSets(statefulSets[i].Namespace.Name).UpdateScale(context.TODO(), statefulSets[i].Statefulset.Name, newScale, metav1.UpdateOptions{})
```

Comment:
StatfulSet `Update` should be happening concurrently similar to the `Create` above.

Reply:
Also, Neha is using deployments in her failover testing. Maybe we should use deployments too to keep the testing consistent.

Reply:
Made the UpdateScale execution concurrent.

Reply:
> 

This is not necessary for us. Lets discuss this offline.